### PR TITLE
Implements polygon SAT collision detection on the overmap [Qwer2D]

### PR DESCRIPTION
**Pull request by Kmc2000**
## About The Pull Request

Thanks to qwerty for the help! If I die it's probably him that did it for butchering this code.

## Why It's Good For The Game

God fuck why overmap hitboxes are so awful NO

## Changelog
:cl: Kmc2000, Qwertyquerty, Monster860
fix: Overmap hitboxes should work properly now.
/:cl:

Co-authored-by: qwertyquerty <qwertytrogi@gmail.com>

### CODE REVIEW

File: nsv13/code/modules/collision/vector2d.dm [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+/*
+Method to turn this vector counter clockwise by a desired angle
+*/
+
+/datum/vector2d/proc/rotate(angle)
+	var/s = sin(angle)
+	var/c = cos(angle)
+	newx = c*x + s*y
+	newy = -s*x + c*y
```
Comment (monster860) [inline nsv13/code/modules/collision/vector2d.dm:32]: wait wtf how does this not error.....

**Comment by Cenrus** [comment]
yes. NSV is saved.

File: nsv13/code/modules/collision/vector2d.dm (hunk 1 above)
Comment (qwertyquerty) [reply nsv13/code/modules/collision/vector2d.dm:32]: It does. This is a draft of a PR. She just put it up so I can review as she writes

**Comment by qwertyquerty** [comment]
image

**Comment by qwertyquerty** [comment]
For more information, see: https://github.com/qwertyquerty/collision

File: nsv13/code/modules/collision/shape.dm [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+/datum/shape/proc/test_aabb(var/datum/shape/other)
+	return  ((src.aabb[1] + src.position.x) <= (other.aabb[3] + other.position.x)) &&
+			((src.aabb[2] + src.position.y) <= (other.aabb[4] + other.position.y)) &&
+			((src.aabb[3] + src.position.x) >= (other.aabb[1] + other.position.x)) &&
+			((src.aabb[4] + src.position.y) >= (other.aabb[2] + other.position.y))
+
+/**
+
+Method to test if two shape objects are colliding with one another. We start off with a relatively inexpensive "Axis Oriented Bounding Box" check to test a block collision. If theyre not colliding with that, then it's safe
```
Comment (qwertyquerty) [inline nsv13/code/modules/collision/shape.dm:112]: Axis Aligned Bounding Box

not oriented

File: nsv13/code/modules/collision/shape.dm [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		src.normals[i].set(edge.perp().normalize())
+	}
+	aabb.Add(new /datum/point(x_max, y_max))
+	aabb.Add(new /datum/point(x_max, y_min))
+	aabb.Add(new /datum/point(x_min, y_min))
+	aabb.Add(new /datum/point(x_min, y_max))
+
+/**
+
+Simple method to calculate whether we collide with another shape object, lightweight but not hugely precise.
```
Comment (qwertyquerty) [inline nsv13/code/modules/collision/shape.dm:100]: Explain that it just checks an AABB collision. Most people know what that means

**Comment by qwertyquerty** [comment]
This allows us to define convex hit boxes like so
image

File: nsv13/code/modules/collision/shape.dm [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+@returns boolean true / false
+
+*/
+
+/datum/shape/proc/collides(var/datum/shape/other)
+	if(!src.test_aabb(other)){
+		return FALSE
+	}
+	for (var/norm in src.normals){
+		if(is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```
Comment (monster860) [inline nsv13/code/modules/collision/shape.dm:132]: we want to return if they *are* colliding, which means *no* separating axis so it should be
```suggestion
		if(!is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```

File: nsv13/code/modules/collision/shape.dm [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+			return FALSE
```
Comment (monster860) [inline nsv13/code/modules/collision/shape.dm:133]:
```suggestion
			return TRUE
```
and if theres no separating axis, then they are colliding

File: nsv13/code/modules/collision/shape.dm [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		}
+	}
+	for (var/norm in other.normals){
+		if(is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```
Comment (monster860) [inline nsv13/code/modules/collision/shape.dm:137]:
```suggestion
		if(!is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```
ditto

File: nsv13/code/modules/collision/shape.dm [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+			return FALSE
```
Comment (monster860) [inline nsv13/code/modules/collision/shape.dm:138]:
```suggestion
			return TRUE
```
ditto

File: nsv13/code/modules/collision/shape.dm [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		}
+	}
+	return TRUE
```
Comment (monster860) [inline nsv13/code/modules/collision/shape.dm:141]:
```suggestion
	return FALSE
```
and if they are all separating axes, then they're not colliding

**Comment by qwertyquerty** [comment]
Monster you can probably just ignore this PR until we tell you to review it. It has a lot coming.

**Comment by Kmc2000** [comment]
I want die

**Comment by qwertyquerty** [comment]
mfw we've been at this for 6 hours straight

**Comment by qwertyquerty** [comment]
7 hours

**Comment by Kmc2000** [comment]
and counting

**Comment by qwertyquerty** [comment]
https://streamable.com/v3ompe

**Comment by Cenrus** [comment]
Hitboxes seem fine apart from syndicate fighters.

File: nsv13/code/modules/overmap/physics.dm [hunk 9]
```diff
@@ -32,25 +39,52 @@
...
+		collider2d = new /datum/shape(position, collision_positions, angle) // -TORADIANS(src.angle-90)
+	else
+		message_admins("[src] does not have collision points set! It will float through everything.")
 
 /obj/structure/overmap/proc/can_move()
 	return TRUE //Placeholder for everything but fighters. We can later extend this if / when we want to code in ship engines.
 
 /obj/structure/overmap/process(time)
-	time /= 10 // fuck off with your deciseconds
+	time /= 1 SECONDS
```
Comment (Jalleo) [inline nsv13/code/modules/overmap/physics.dm:82]: What the HECK IS THIS DOING HERE!


Okay after this PR YOU MUST REWRITE ALL OF THIS TO PROPERLY USE A SUBSYSTEM.

File: nsv13/code/modules/collision/vector2d.dm [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+/datum/vector2d
+	var/x
+	var/y
+
+/*
+Constructor for vector2d objects, taking a simple X,Y coordinate.
+*/
+/datum/vector2d/New(x = 0, y = 0)
+	src.x = x
```
[the remark this answers is not in the document]
Comment (Kmc2000) [reply nsv13/code/modules/collision/vector2d.dm:19]: No. This is the equivalent of this.x = x. Name overloading.

File: nsv13/code/modules/collision/vector2d.dm [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+*/
+/datum/vector2d/proc/angle()
+	return ATAN2(src.x, src.y)
+
+/*
+Normalize the vector so it has a magnitude of 1
+@return a normalized version of the vector
+*/
+/datum/vector2d/proc/normalize()
+	RETURN_TYPE(/datum/vector2d)
```
[the remark this answers is not in the document]
Comment (Kmc2000) [reply nsv13/code/modules/collision/vector2d.dm:159]: It's for spacemanDMM and is the equivalent of something like public Vector2d returnVector, or public static int getNum

File: nsv13/code/modules/overmap/physics.dm (hunk 9 above)
Comment (Kmc2000) [reply nsv13/code/modules/overmap/physics.dm:82]: ...?
This process() is evidence that it already does use a subsystem, overmap processing.

File: nsv13/code/modules/collision/shape.dm [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+	var/list/datum/vector2d/normals = list()
+	var/list/aabb = list() //Cached points from AABB collision
+
+//Constructor for shape objects, taking in parameters they may initially need
+
+/datum/shape/New(datum/vector2d/position, list/points, _angle=0)
+	. = ..()
+	if(!position)
+		position = new /datum/vector2d()
+	src.position = position
```
[the remark this answers is not in the document]
Comment (Kmc2000) [reply nsv13/code/modules/collision/shape.dm:27]: Name overload. Sets this.position to a new positional vector.

**Pull request merged**